Driver stm32 wl dev functional

ludoch-stm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

jeromecoutant committed:
STM32WL: STM32Cube_FW_WL_V1.0.0

jeromecoutant committed:
STM32WL: update for MBED

jeromecoutant committed:
STM32WL: NUCLEO_WL55JC creation

jeromecoutant committed:
STM32WL: targets.json update

JeanMarcR committed:
INTEGRATION OF NUCLEO_WL55JC IN MBED

Jerome

ludoch-stm committed:
STM32WL dir created: 1st build, to be tested

ludoch-stm committed:
STM32WL driver adapted: to be tested

jeromecoutant committed:
STM32WL: enable uvision6 export

ludoch-stm committed:
STM32WL driver: add target in mbed_lib.json for export

ludoch-stm committed:
STM32WL: fix driver build (still one build error)

jeromecoutant committed:
STM32WL55JCIx: cmsis pack update

jeromecoutant committed:
Quick patch for compilation

jeromecoutant committed:
Patch from https://github.com/ARMmbed/mbed-os/pull/10424

jeromecoutant committed:
STM32WL55JCIx: cmsis pack update again

jeromecoutant committed:
STM32WL55JCIx Keil download ok

ludoch-stm committed:
interface change and callbacks defined: link KO

jeromecoutant committed:
WORKAROUND: disable deepsleep

ludoch-stm committed:
fix interface change + add NVIC registering of HW IT: SUBGHZ_Radio_IRQn: Keil download OK

ludoch-stm committed:
add MBED_CONF_RTOS_PRESENT possibility (still not functional, SUBGHZ_Radio_IRQHandler not triggered)

ludoch-stm committed:
fix: call to SUBGHZ_Radio_IRQn

ludoch-stm committed:
fix: call to HAL_SUBGHZ_ callbacks from C++ to C (running but not functional in front of GW)

ludoch-stm committed:
feature: add RFswitch configuration setup => Join to Network Server

ludoch-stm committed:
fix: crystal_select param handling + cleanup

ludoch-stm committed:
feature: add debug mode, debug pins

ludoch-stm committed:
fix: include radio board in mbed env + remove debug files

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+/**
+  * @brief Set RX pin to high or low level
+  */
+#define DBG_GPIO_RADIO_RX(set_rst) DBG_GPIO_##set_rst##_LINE(DGB_LINE1_PORT, DGB_LINE1_PIN);
```
Are those GPIOs always available for debug ?
Shouldn't this be under compilation switch ?

ludoch-stm replied:
I have removed this def in the last version, and intend to set these Dbg pins as input class member. 
They are commented for now, because I still have issues due to static function including them.

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+#if MBED_CONF_STM32WL_LORA_DRIVER_CRYSTAL_SELECT== 1
+  uint8_t _crystal_select  = 1;
+#else
+	uint8_t _crystal_select = 0;
```
indentation doesn't seem ok here - maybe run astyle ? check with @jeromecoutant  how to proceed

ludoch-stm replied:
I have run Astyle on last pull request

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+ 
+// Data buffer used for both TX and RX
+// Size of this buffer is configurable via Mbed config system
+// Default is 255 bytes
```
It could be interesting to describe when buffer needs to be increased or decreased .... depends on use case or on network protocol ?

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+static uint8_t _active_modem;
+    
+using namespace mbed;
+using namespace rtos;
```
Shouldn't this be below line 89 ? if MBED_CONF_RTOS_PRESENT is not present I don't think that namespace rtos would exist ? 
actually do we plan to support baremetal as well ?

LMESTM replied:
Also if both rtos based and bare metal profiles are supported, it would be great to briefly explain advantage of the rtos and the differences in behavior.

ludoch-stm replied:
using namespace rtos; set under #define MBED_CONF_RTOS_PRESENT: done
Concerning Baremetal, this is a good question to be discussed together. I haven't looked at it yet

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+
+        lock();
+        if (flags & SIG_INTERRUPT) {
+            irq_process_isr();
```
Not clear to me how this works ?
when flags has SIG_INTERRUPT you will call irq_process_isr that then set the flags_set(SIG_INTERRUPT);
Looks like a kind of deadlock ....
don't you rather want to call RadioIrqProcess here ?

ludoch-stm replied:
You are right, strange the deadlock didn't occur: corrected

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+    write_opmode_command((uint8_t) RADIO_CLR_IRQSTATUS, buf, 2);
+}
+
+//TODO - Better do CAD here. CAD code is already part of the driver
```
TODO to be handled

ludoch-stm replied:
comment from sx126x radio implementation: removed

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+    receive();
+
+    // hold on a bit, radio turn-around time
+    ThisThread::sleep_for(1);
```
Do we wait for 1ms or 1s.
this can be made explicit by using: ThisThread::sleep_for(1ms);

ludoch-stm replied:
OK: done

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+  */
+static void SUBGHZ_Radio_IRQHandler(void)
+{
+  /* USER CODE BEGIN SUBGHZ_Radio_IRQn 0 */
```
do we want to keep the user sections generated from MX. Can be kept if this will be useful for later maintenance and in case the code is copy-paster directly from MX

ludoch-stm replied:
Better to remove them, not useful in mbed-os env: done

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+//    ThisThread::sleep_for(2);
+//    _reset_ctl.input();
+    // give some time for automatic image calibration
+    ThisThread::sleep_for(6);
```
proposal to use 6ms

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+    write_opmode_command(RADIO_SET_SLEEP, &sleep_state, 1);
+  
+    _operation_mode = MODE_SLEEP;
+    ThisThread::sleep_for(2);
```
2ms
Is it required to wait 2ms ?
in mbed-os the deep sleep mode can be entered / exited quite frequencly, as soon as there is nothing to do for 10ms. so 2ms is quite a lot, and any saving will end up in power saving (this comment is valid for all the timings that happen very frequencly)

ludoch-stm replied:
yes, this delay is present in sx126x radio Semtech implementation. 
Not sure how 2ms is calculated, because in sx126x datasheet, it's explained that  after SetSleep command, device will become unresponsive for around 0,5ms

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/radio_board_if.h:
```diff
@@ -0,0 +1,238 @@
+/* In order to use BSP driver, add the correspondent files in the IDE workspace */
+/* and define USE_BSP_DRIVER in the preprocessor definitions  or in platform.h */
+
+#elif defined(MX_NUCLEO_WL55JC1)
```
@JeanMarcR  @jeromecoutant  We should make sure that those control pins are not offered for other preipherals / purpose in our mbed pinmap !

ludoch-stm replied:
file has been removed, pins are now defined in mbed_app.json

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+    SUBGRF_SetTxParams(RFO_LP, 0, RADIO_RAMP_200_US);
+	
+     /* Configure the debug mode*/
+    DBG_Init();
```
Do we always want to have DBG mode ? How is it configured ?

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/sys_conf.h:
```diff
@@ -0,0 +1,72 @@
+  * @brief Enable Debugger mode
+  * @note  1:ON it enables the debbugger plus 4 dgb pins, 0:OFF the debugger is OFF (lower consumption)
+  */
+#define DEBUGGER_ON       0
```
Could be nice to offer this option from the mbed_app.json

LMESTM reviewed (commented):
Thanks @ludoch-stm for all the good work 👍 
Several comments or questions shared ... as I'm not expert on LORA so those are more generic comments.

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/SleepMode.txt:
```diff
@@ -0,0 +1,22 @@
+Sleep Modes:
+
+The SX126X series LoRa radios define two different sleep modes, namely:
```
SX126X series radios : Must be replaced by STM32WL Subghz radio

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/SleepMode.txt:
```diff
@@ -0,0 +1,22 @@
+
+Sleep mode with Cold Start:
+The driver can be configured to sleep with cold startup. This mode is the lowest power consuming state 
+for the SX126X series radios. No configurations are retained in this mode, that's why our driver takes 
```
SX126X series radios : Must be replaced by STM32WL Subghz radio

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WL55xC/TOOLCHAIN_GCC_ARM/startup_stm32wl55xx.S:
```diff
@@ -0,0 +1,434 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wl55xx_cm4.s
```
align filename and this field. Also replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WL55xC/TOOLCHAIN_ARM/startup_stm32wl55xx.S:
```diff
@@ -0,0 +1,320 @@
+;******************************************************************************
+;* File Name          : startup_stm32wl55xx_cm4.s
```
align filename and this field. Also replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WL55xC/TOOLCHAIN_IAR/startup_stm32wl55xx.S:
```diff
@@ -0,0 +1,509 @@
+;********************************************************************************
+;* File Name          : startup_stm32wl55xx_cm4.s
```
align filename and this field. Also replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4x8/TOOLCHAIN_ARM/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,314 @@
+;******************************************************************************
+;* File Name          : startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4x8/TOOLCHAIN_GCC_ARM/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,425 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4x8/TOOLCHAIN_IAR/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,494 @@
+;********************************************************************************
+;* File Name          : startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4xB/TOOLCHAIN_ARM/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,314 @@
+;******************************************************************************
+;* File Name          : startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4xB/TOOLCHAIN_GCC_ARM/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,425 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4xB/TOOLCHAIN_IAR/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,494 @@
+;********************************************************************************
+;* File Name          : startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4xC/TOOLCHAIN_ARM/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,314 @@
+;******************************************************************************
+;* File Name          : startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4xC/TOOLCHAIN_GCC_ARM/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,425 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE4xC/TOOLCHAIN_IAR/startup_stm32wle4xx.S:
```diff
@@ -0,0 +1,494 @@
+;********************************************************************************
+;* File Name          : startup_stm32wle4xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5x8/TOOLCHAIN_ARM/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,314 @@
+;******************************************************************************
+;* File Name          : startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5x8/TOOLCHAIN_GCC_ARM/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,425 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5x8/TOOLCHAIN_IAR/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,494 @@
+;********************************************************************************
+;* File Name          : startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5xB/TOOLCHAIN_ARM/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,314 @@
+;******************************************************************************
+;* File Name          : startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5xB/TOOLCHAIN_GCC_ARM/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,425 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5xB/TOOLCHAIN_IAR/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,494 @@
+;********************************************************************************
+;* File Name          : startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5xC/TOOLCHAIN_IAR/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,494 @@
+;********************************************************************************
+;* File Name          : startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5xC/TOOLCHAIN_GCC_ARM/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,425 @@
+/**
+  ******************************************************************************
+  * @file      startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on targets/TARGET_STM/TARGET_STM32WL/TARGET_STM32WLE5xC/TOOLCHAIN_ARM/startup_stm32wle5xx.S:
```diff
@@ -0,0 +1,314 @@
+;******************************************************************************
+;* File Name          : startup_stm32wle5xx.s
```
replace the .S uppercase of file

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+
+void STM32WL_LoRaRadio::HAL_SUBGHZ_CRCErrorCallback (void)
+{
+   if (_radio_events && _radio_events->rx_error) 
```
why do your check "_radio_events" only in this callback and not in other functions ?

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+            state = RBI_SWITCH_RFO_LP;
+            Radio_SMPS_Set(SMPS_DRIVE_SETTING_MAX);
+        }
+        if (paSelect == RFO_HP)
```
"if..else" maybe, instead of "if..if" ? paSelect have only 2 values.

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+        _image_calibrated = true;
+    }
+
+    freq = (uint32_t) ceil(((float) frequency / (float) FREQ_STEP));
```
remove extra parenthesis 
ceil((float) frequency / (float) FREQ_STEP)

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+#if MBED_CONF_STM32WL_LORA_DRIVER_STANDBY_MODE == 1
+    uint8_t standby_mode = 1;
+#else
+	uint8_t standby_mode = 0;
```
fix indentation (more similar changes on next files. please run astyle tools)

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1561 @@
+    cold_start_wakeup();
+    
+    SUBGRF_SetTxParams(RFO_LP, 0, RADIO_RAMP_200_US);
+	
```
missing ? SUBGRF_SetDioIrqParams( IRQ_RADIO_ALL, IRQ_RADIO_ALL, IRQ_RADIO_NONE, IRQ_RADIO_NONE );

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.h:
```diff
@@ -0,0 +1,412 @@
+  *
+  *          Portions COPYRIGHT 2020 STMicroelectronics
+  *
+  * @file    STM32WL_loRaRadio.h
```
align file field on filename (case sensitive)

ludoch-stm replied:
Ok, done

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.h:
```diff
@@ -0,0 +1,412 @@
+    void SUBGRF_SetTxParams( uint8_t paSelect, int8_t power, radio_ramp_time_t rampTime );
+    void Radio_SMPS_Set(uint8_t level);
+    
+private:
```
already in private section. remove duplicate

ludoch-stm replied:
Ok, done

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_radio_driver.h:
```diff
@@ -0,0 +1,676 @@
+
+
+/*!
+ * \brief List of matching supported by the sx126x
```
replace sx126x by STM32WL Subghz

PF-STM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_radio_driver.h:
```diff
@@ -0,0 +1,676 @@
+/*!
+ * \file      STM32WL_radio_driver.h
+ *
+ * \brief     STM32WL driver implementation
```
STM32WL **radio** driver implementation

ludoch-stm replied:
Ok done

jeromecoutant replied:
Are you sure it's done ? :-)

PF-STM commented on connectivity/lorawan/mbed_lib.json:
```diff
@@ -93,5 +93,10 @@
-    }
+    },
+	"target_overrides": {
+          "NUCLEO_WL55RC": {
```
6 features of STM32WL are available:
Dual Core: STM32WL5xCx, STM32WL5xJx and STM32WL5xUx
Single Core: STM32WLE5Cx, STM32WLE5Jx and STM32WLE5Ux

But it seems that WL55RC is not existing in this reference list.

jeromecoutant replied:
yes, correct name is NUCLEO_WL55JC :-)
But if this wakeup time constraint is valid for more than this NUCLEO board, use STM32WL55xC or even STM32WL?

ludoch-stm committed:
Run Astyle + Codespell

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+
+Maintainer: Miguel Luis, Gregory Cristian & Gilbert Menth
+
+Copyright (c) 2019, Arm Limited and affiliates.
```
This Semtech header is quite strange  for me...
And there is only ARM copyright...
No Semtech, no ST ?

LMESTM replied:
I had same question at first glance but actually semtech copyright is line 8. and ST one line 28. 
Overall looks ok to me

ludoch-stm replied:
As this file has mainly been designed by Semtech legals told us to keep it, for WL project, and as we made some specific  changes for our radio, we add the portions with our ST Copyright

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+#include "stm32wlxx_hal_subghz.h"
+
+
+#if MBED_CONF_STM32WL_LORA_DRIVER_REGULATOR_MODE == 1
+uint8_t regulator_mode = USE_DCDC;
+#else
+uint8_t regulator_mode = USE_LDO;
+#endif
```
Minor comment: You could use
uint8_t regulator_mode = MBED_CONF_STM32WL_LORA_DRIVER_REGULATOR_MODE
if you set USE_DCDC or USE_LDO in the json file

ludoch-stm replied:
OK, done

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    _rf_dbg_rx(rf_dbg_rx, PIN_OUTPUT, PullNone, 0),
+    _rf_dbg_tx(rf_dbg_tx, PIN_OUTPUT, PullNone, 0)
+#ifdef MBED_CONF_RTOS_PRESENT
+    , irq_thread(osPriorityRealtime, 1024, NULL, "Thread_STM32WL")
```
Same question as Laurent : can Lora work without thread ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+void STM32WL_LoRaRadio::error_handler(HAL_StatusTypeDef error)
+{
+    printf("ERROR in STM32WL_loRaRadio.c : %d", error);
+    /* USER CODE BEGIN error_handler_Debug */
```
This kind of comment are not useful ?

ludoch-stm replied:
yes, comment removed

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    if (error_value != HAL_OK)
+    {
+        error_handler(error_value);
+    }
```
Nice MBED way of coding could be:
MBED_ASSERT(error_value == HAL_OK);

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+
+    SUBGRF_SetTxParams(RFO_LP, 0, RADIO_RAMP_200_US);
+
+    /* ST_WORKAROUND_BEGIN: Sleep radio */
```
ST_WORKAROUND_BEGIN ??

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    }
+}
+
+void STM32WL_LoRaRadio::sleep(void)
```
Change name into radio_sleep for ex ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    Radio_SMPS_Set(SMPS_DRIVE_SETTING_DEFAULT);
+
+
+    //DBG LCH:check value
```
comment to remove/update ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    ThisThread::sleep_for(2);
+}
+
+uint32_t STM32WL_LoRaRadio::random(void)
```
random procedure name maybe to replace into radio_get_random ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+
+void STM32WL_LoRaRadio::start_cad(void)
+{
+    // TODO: CAD is more advanced in SX126X. We will need API change in LoRaRadio
```
TODO ?

ludoch-stm replied:
removed

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    }
+    else // rfo_hp
+    {
+        // WORKAROUND - Better Resistance of the SX1262 Tx to Antenna Mismatch, see DS_SX1261-2_V1.2 datasheet chapter 15.2
```
workaround ?

ludoch-stm replied:
same comment present on WL BSP: better remove it or let a trace to explain it?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+                      IRQ_RADIO_NONE,
+                      IRQ_RADIO_NONE);
+
+    /* ST_WORKAROUND_BEGIN : Set the debug pin and update the radio switch */
+    /* Set DBG pin */
+//    DBG_GPIO_RADIO_TX(SET);
+//  STM32WL_LoRaRadio::setTXPin(1);
```
to remove ?

ludoch-stm replied:
I still have to find a way to set these debug pins (for now, static parameter compil issue).
I keep them as comments until resolved

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1720 @@
+    }
+
+
+    /* ST_WORKAROUND_BEGIN : Set the debug pin and update the radio switch */
+    /* Set DBG pin */
+//    DBG_GPIO_RADIO_RX(SET);
+//    STM32WL_LoRaRadio::setRXPin(1);
```
to remove ?

ludoch-stm committed:
fix: GitHub code review

ludoch-stm committed:
fix: RX/TX debug pins implementation
fix: code review step2

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -734,26 +716,21 @@ void STM32WL_LoRaRadio::init_radio(radio_events_t *events)
-
-    /* ST_WORKAROUND_BEGIN: Sleep radio */
+    
     sleep();
```
@ludoch-stm  one question: where is this function sleep defined ? Could there be a more precise name, like lora_radio_sleep because sleep is usually also used at system level and should not be called by a driver ?

jeromecoutant replied:
agree, add something like // request radio to sleep ?

ludoch-stm commented:
@LMESTM
This was already a comment of the code review, but unfortunately sleep function is defined in LoraRadio.h, and changing this function name would impact all Lora implementations/targets. As there could add unseen regressions it has been decided to stay with it, although I'm in line with your comment.

LMESTM commented:
> @LMESTM
> This was already a comment of the code review, but unfortunately sleep function is defined in LoraRadio.h, and changing this function name would impact all Lora implementations/targets. As there could add unseen regressions it has been decided to stay with it, although I'm in line with your comment.

ok clear thanks - maybe make the comment even more explicit at each call of sleep() function ?

ludoch-stm committed:
fix: debug RX/TX pins not executed

ludoch-stm committed:
fix: RTOS event set in SUBGHZ_Radio_IRQHandler triggering RadioIrqProcess

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_radio_driver.h:
```diff
@@ -0,0 +1,697 @@
+  * @note   The sub-GHz radio peripheral registers can be accessed by sub-GHz radio command
+  *         SUBGRF_WriteRegisters() and SUBGRF_ReadRegisters() "
+  */
+/*Sub-GHz radio generic bit synchronization register*/
+#define SUBGHZ_GBSYNCR                              REG_BIT_SYNC
+/*Sub-GHz radio generic packet control 1A register*/
+#define SUBGHZ_GPKTCTL1AR                           REG_LR_WHITSEEDBASEADDR_MSB
+/*Sub-GHz radio generic whitening LSB register*/
+#define SUBGHZ_GWHITEINIRL                          REG_LR_WHITSEEDBASEADDR_LSB  
+/*Sub-GHz radio generic CRC initial MSB register*/
+#define SUBGHZ_GCRCINIRH                            REG_LR_CRCSEEDBASEADDR
+/*Sub-GHz radio generic CRC initial LSB register*/
+#define SUBGHZ_GCRCINIRL                            0x06BD
+/*Sub-GHz radio generic CRC polynomial MSB register*/
+#define SUBGHZ_GCRCPOLRH                            REG_LR_CRCPOLYBASEADDR
+/*Sub-GHz radio generic CRC polynomial LSB register*/
+#define SUBGHZ_GCRCPOLRL                            0x06BF
+/*Sub-GHz radio generic synchronization word control register 7*/
+#define SUBGHZ_GSYNCR7                              REG_LR_SYNCWORDBASEADDRESS
+/*Sub-GHz radio generic synchronization word control register 6*/
+#define SUBGHZ_GSYNCR6                              0x06C1
+/*Sub-GHz radio generic synchronization word control register 5*/
+#define SUBGHZ_GSYNCR5                              0x06C2 
+/*Sub-GHz radio generic synchronization word control register 4*/
+#define SUBGHZ_GSYNCR4                              0x06C3
+/*Sub-GHz radio generic synchronization word control register 3*/
```
Why all these register definitions are not defined in CMSIS file like other registers ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+  */
+
+
+#include <math.h>
+#include "ThisThread.h"
+#include "mbed_wait_api.h"
+#include "Timer.h"
+#include "STM32WL_LoRaRadio.h"
+#include "stm32wlxx_hal_subghz.h"
```
seems that only 
#include <math.h>
#include "Timer.h"
#include "STM32WL_LoRaRadio.h"
is useful ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+  *
+  *          Portions COPYRIGHT 2021 STMicroelectronics
+  *
+  * @file    STM32WL_LoRaRadio.c
```
cpp :-)

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+const uint8_t sync_word[] = {0xC1, 0x94, 0xC1, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+// in ms                                 SF12    SF11    SF10    SF9    SF8    SF7
+const float lora_symbol_time[3][6] = {{ 32.768, 16.384, 8.192, 4.096, 2.048, 1.024 },  // 125 KHz
+    { 16.384, 8.192,  4.096, 2.048, 1.024, 0.512 },  // 250 KHz
+    { 8.192,  4.096,  2.048, 1.024, 0.512, 0.256 }
+}; // 500 KHz
```
minor, but maybe need alignment ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+    }
+    // ERROR: Value not found
+    // This should never happen
+    while (1);
```
Maybe call MBED_ASSERT(0) ?

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+
+    /* Set RF switch */
+    SUBGRF_SetSwitch(_antSwitchPaSelect, RFSWITCH_TX);
+    /* ST_WORKAROUND_END */
```
to remove

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+    
+    /* RF switch configuration */
+    SUBGRF_SetSwitch(_antSwitchPaSelect, RFSWITCH_RX);
+    /* ST_WORKAROUND_END */
```
to remove

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.h:
```diff
@@ -0,0 +1,405 @@
+/**
+  ******************************************************************************
+  *
+  *          Portions COPYRIGHT 2020 STMicroelectronics
```
2021

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.h:
```diff
@@ -0,0 +1,405 @@
+#define MBED_LORA_RADIO_DRV_STM32WL_LORARADIO_H_
+
+#include "mbed_critical.h"
+#include "PinNames.h"
+#include "InterruptIn.h"
+#include "DigitalOut.h"
+#include "DigitalInOut.h"
+#include "DigitalIn.h"
+#include "AnalogIn.h"
```
Only:
#include "mbed_critical.h"
#include "DigitalOut.h"
#include "platform/PlatformMutex.h"

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.h:
```diff
@@ -0,0 +1,405 @@
+private:
+
+    // Radio specific controls (TX/RX duplexer switch control)
+    mbed::DigitalInOut _rf_switch_ctrl1;
+    mbed::DigitalInOut _rf_switch_ctrl2;
+    mbed::DigitalInOut _rf_switch_ctrl3;
```
Pins seems never used in In mode,
so use DigitalOut type

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+
+
+#ifdef MBED_CONF_STM32WL_LORA_DRIVER_DEBUG_RX
+static DigitalInOut _rf_dbg_rx(MBED_CONF_STM32WL_LORA_DRIVER_DEBUG_RX, PIN_OUTPUT, PullNone, 0);
```
```suggestion
static DigitalOut _rf_dbg_rx(MBED_CONF_STM32WL_LORA_DRIVER_DEBUG_RX, 0);
```

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+#endif
+
+#ifdef MBED_CONF_STM32WL_LORA_DRIVER_DEBUG_TX
+static DigitalInOut _rf_dbg_tx(MBED_CONF_STM32WL_LORA_DRIVER_DEBUG_TX, PIN_OUTPUT, PullNone, 0);
```
```suggestion
static DigitalOut _rf_dbg_tx(MBED_CONF_STM32WL_LORA_DRIVER_DEBUG_TX, 0);
```

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -0,0 +1,1666 @@
+                                     PinName rf_switch_ctrl3
+                                    )
+    :
+    _rf_switch_ctrl1(rf_switch_ctrl1, PIN_OUTPUT, PullNone, 0),
+    _rf_switch_ctrl2(rf_switch_ctrl2, PIN_OUTPUT, PullNone, 0),
+    _rf_switch_ctrl3(rf_switch_ctrl3, PIN_OUTPUT, PullNone, 0)
```
```suggestion
    _rf_switch_ctrl1(rf_switch_ctrl1, 0),
    _rf_switch_ctrl2(rf_switch_ctrl2, 0),
    _rf_switch_ctrl3(rf_switch_ctrl3, 0)
```

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.h:
```diff
@@ -0,0 +1,405 @@
+    // helper functions
+    void wakeup();
+
+    // void set_dio2_as_rfswitch_ctrl(uint8_t enable);
```
remove this line

jeromecoutant commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_radio_driver.h:
```diff
@@ -0,0 +1,697 @@
+/*!
+ * \file      STM32WL_radio_driver.h
```
Miss license header

jeromecoutant reviewed (changes requested):
Appli is crashing with the last commit...

@LMESTM could you have a look ?

jeromecoutant committed:
REVIEW

LMESTM commented on connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp:
```diff
@@ -202,6 +206,22 @@ void STM32WL_LoRaRadio::rf_irq_task(void)
+  
+    irq_thread.flags_set(SIG_INTERRUPT);
+  
+    irq_status_rtos = (radio_irq_masks_t)STM32WL_LoRaRadio::get_irq_status();
```
Even in case of using a thread, I would tend to keep reading this status only in one place : in RadioIrqProcess. Otherwise there is a risk of this value is modified in this IRQ handler, while it is being use din the thread.
So either only access it in one place (my first recommendation) or protect it against multiple access

LMESTM replied:
> @LMESTM could you have a look ?

I had a look and don't see reasons for a crash. But I think there is a possible race condition that should be avoided so I added a comment for @ludoch-stm to look at

jeromecoutant committed:
Baremetal support

Added line in mbed_app.json:
"requires": ["bare-metal", "lora", "mbedtls", "events", "stm32wl-lora-driver"],

ludoch-stm commented:
@LMESTM When implementing the read status and clear irq in RadioIrqProcess, the system goes in :" CMSIS-RTOS error: ISR Queue overflow", due to the fact that irq isn't cleared and so program loops indefinitely in rf_irq_task. That's the reason why I separated it.
Concerning the protection against multiple access, you're rigth, it should be done.
I added core_util_critical_section_enter()/exit(), in RadioIrqProcess around  irq_status = irq_status_rtos;
but mutex error remains unchanged

ludoch-stm committed:
fix: remove irq thread implementation, RadioIRQ treated at IT level in LoRaRadio driver
